### PR TITLE
[SYCL] Fix multi_ptr relational operators customised for nullptr.

### PR DESCRIPTION
**Pull request by mmoadeli**
`multi_ptr` relational operators taking a `std::nullptr_t` are written in a way that assume it is the lowest possible value
(example https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/multi_ptr.hpp#L1575). However the C++ specs states there is no ordering requirement (https://eel.is/c++draft/expr.rel#4.3).

In practice, this is causing issues in the CUDA and AMDGPU backend. For instance, in CUDA the nullptr in the local address space is a non `0` value and `0` in this address space is the root of the allocated local memory.

### CODE REVIEW

File: sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+#include <sycl/sycl.hpp>
```
Comment (aelovikov-intel) [inline sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp:4]: How hard would it be to switch it to `<sycl/detail/core.hpp>` + something? I'd hope

```suggestion
#include <sycl/detail/core.hpp>
#include <sycl/multi_ptr.hpp>
```

should be enough.

File: sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+template <typename multi_ptr_t> void nullptrRelationalOperatorTest() {
+  constexpr int OUTPUT_SIZE = 8;
+  bool output[OUTPUT_SIZE];
+  try {
+    sycl::queue queue;
+    sycl::buffer<bool, 1> buf(output, sycl::range<1>(OUTPUT_SIZE));
```
Comment (aelovikov-intel) [inline sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp:11]: nit, my preference would be

```
  sycl::buffer<bool, 1> buf(N);
  /* do work */
  for (bool res : sycl::host_accessor{buf})
    assert(res);
```

until we will have ranges and `all_of` on a range.

File: sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    queue
+        .submit([&](sycl::handler &cgh) {
+          auto dev_acc = buf.get_access<sycl::access::mode::write>(cgh);
+          sycl::local_accessor<bool, 1> locAcc(1, cgh);
+          cgh.parallel_for(sycl::nd_range<1>{1, 1}, [=](sycl::id<1>) {
+            locAcc[0] = 10;
+            multi_ptr_t mp(locAcc);
+            dev_acc[0] = std::less<multi_ptr_t>()(nullptr, mp) == nullptr < mp;
```
Comment (aelovikov-intel) [inline sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp:19]: Should we also check the result using plain decorated pointers without `multi_ptr`?

File: sycl/include/sycl/multi_ptr.hpp [hunk 4]
```diff
@@ -1546,70 +1546,80 @@ template <typename ElementType, access::address_space Space,
           access::decorated DecorateAddress>
 bool operator==(const multi_ptr<ElementType, Space, DecorateAddress> &lhs,
                 std::nullptr_t) {
-  return lhs.get() == nullptr;
+  return lhs.get() ==
+         multi_ptr<ElementType, Space, DecorateAddress>(nullptr).get();
```
Comment (Naghasan) [inline sycl/include/sycl/multi_ptr.hpp:1550]: I don't understand that change here. With sycl tip, the following

```cpp
__attribute__((sycl_device)) bool f(__attribute__((opencl_local))int* i) {
    return i == nullptr;
}
```

generates this for nvptx

```llvm
%cmp = icmp eq ptr addrspace(3) %i, addrspacecast (ptr null to ptr addrspace(3))
```

which is what we expect. If you have an issue `lhs.get() == nullptr`, this probably hides a compiler issue we need logged.

Comment (mmoadeli) [reply sycl/include/sycl/multi_ptr.hpp:1550]: thanks @Naghasan 
You are right. There was no need for this change for `==`. For other operators having something like `lhs.get() > nullptr` results in build error.

Comment (mmoadeli) [reply sycl/include/sycl/multi_ptr.hpp:1550]: Thanks @Naghasan 
You are right  regarding `==` operator. It fails to build for other relational operators such as `lhs.get() > nullptr`.

File: sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp (hunk 3 above)
Comment (mmoadeli) [reply sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp:19]: I'm not quite sure I understand what you mean by "decorated pointers without multiptr." Could you please provide more context?

Comment (aelovikov-intel) [reply sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp]: `decltype(multi_ptr_obj.get_decorated())` I would be something like `int __attribute__((opencl)) *`.

Comment (mmoadeli) [reply sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp]: @aelovikov-intel Testing plain decorated pointers is beyond the scope of this PR, as it exclusively addresses relational operators within multi_ptrs.

File: sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp (hunk 1 above)
Comment (mmoadeli) [reply sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp]: thanks @aelovikov-intel, done.

File: sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp (hunk 2 above)
Comment (mmoadeli) [reply sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp]: thanks @aelovikov-intel, done.

File: sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  }
+}
+
+int main() {
+  using multi_ptr_yes =
+      sycl::multi_ptr<bool, sycl::access::address_space::local_space,
+                      sycl::access::decorated::yes>;
+  using multi_ptr_no =
+      sycl::multi_ptr<bool, sycl::access::address_space::local_space,
+                      sycl::access::decorated::no>;
```
[the remark this answers is not in the document]
Comment (mmoadeli) [reply sycl/test-e2e/Basic/multi_ptr_null_relational_operators.cpp]: thanks @Naghasan , done.

File: sycl/include/sycl/multi_ptr.hpp (hunk 4 above)
Comment (mmoadeli) [reply sycl/include/sycl/multi_ptr.hpp]: thanks @Naghasan , that particular change is reverted.

**Review by aelovikov-intel — APPROVED** [review]

**Comment by aarongreig** [comment]
I think this is causing an issue in CI
https://github.com/intel/llvm/actions/runs/8539049320/job/23393549122?pr=13041#step:21:2175
https://github.com/intel/llvm/actions/runs/8538371659/job/23391647615#step:21:2178
https://github.com/intel/llvm/actions/runs/8537393047/job/23388301607#step:21:2160

**Comment by AlexeySachkov** [comment]
> I think this is causing an issue in CI https://github.com/intel/llvm/actions/runs/8539049320/job/23393549122?pr=13041#step:21:2175 https://github.com/intel/llvm/actions/runs/8538371659/job/23391647615#step:21:2178 https://github.com/intel/llvm/actions/runs/8537393047/job/23388301607#step:21:2160

The fix is on review here: #13268

**Pull request merged**